crates/virtio-queue/src/mock.rs

```diff
@@ -0,0 +1,327 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Mocking
+
+use std::marker::PhantomData;
+use std::mem::size_of;
+use std::ops::Index;
+
+use vm_memory::{Address, ByteValued, Bytes, GuestAddress, GuestAddressSpace, GuestMemory, GuestUsize};
+
+use crate::defs::{VIRTQ_DESC_F_INDIRECT, VIRTQ_DESC_F_NEXT};
+use crate::Queue;
+
+pub struct Ref<'a, M, T> {
+    mem: &'a M,
+    addr: GuestAddress,
+    phantom: PhantomData<*const T>,
+}
+
+impl<'a, M: GuestMemory, T: ByteValued> Ref<'a, M, T> {
+    fn new(mem: &'a M, addr: GuestAddress) -> Self {
+        Ref {
+            mem,
+            addr,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn load(&self) -> T {
+        self.mem.read_obj(self.addr).unwrap()
+    }
+
+    pub fn store(&self, val: T) {
+        self.mem.write_obj(val, self.addr).unwrap()
+    }
+}
+
+pub struct ArrayRef<'a, M, T> {
+    mem: &'a M,
+    addr: GuestAddress,
+    len: usize,
+    phantom: PhantomData<*const T>,
+}
+
+impl<'a, M: GuestMemory, T: ByteValued> ArrayRef<'a, M, T> {
+    fn new(mem: &'a M, addr: GuestAddress, len: usize) -> Self {
+        ArrayRef {
+            mem,
+            addr,
+            len,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn ref_at(&self, index: usize) -> Ref<'a, M, T> {
+        // TODO: add better error handling to the mock logic.
+        assert!(index < self.len);
+
+        let addr = self
+            .addr
+            .checked_add((index * size_of::<T>()) as u64)
+            .unwrap();
+
+        Ref::new(self.mem, addr)
+    }
+}
+
+// impl<'a, M: GuestMemory, T: ByteValued> Index<usize> for ArrayRef<'a, M, T> {
```

Comment:
We can delete the commented stuff if we don't end up using it later. In particular, the index operator seems like it could have been nice here as a more concise way of calling `ref_at`. Is the commented implementation not working?